app/assets/javascripts/admin/modules/analytics.js

```diff
@@ -1,13 +1,26 @@
 (function (GOVUK) {
   'use strict'
 
+  GOVUK.setCustomDimensionsFromMetaTags = function () {
+    var metas = document.querySelectorAll("meta[name^='custom-dimension']")
+
+    for (var i = 0; i < metas.length; i++) {
+      var meta = metas[i]
+      var dimensionId = parseInt(meta.getAttribute('name').split('custom-dimension:')[1])
+      var content = meta.getAttribute('content')
+
+      GOVUK.analytics.setDimension(dimensionId, content)
+    }
+  }
+
   GOVUK.Analytics.load()
   GOVUK.analyticsVars = { primaryLinkedDomains: [document.domain] }
   GOVUK.analytics = new GOVUK.Analytics({
     universalId: 'UA-26179049-6', // GOVUK Apps GA ID
     cookieDomain: document.domain
   })
 
+  GOVUK.setCustomDimensionsFromMetaTags()
```

Comment:
Is there a reason why this is attached to the GOVUK namespace?

Reply:
I was having trouble accessing the function in the test file, attaching it to GOVUK allowed me to call it when testing. I wasn't sure if there's a better/cleaner way to do this.

Reply:
This seems like a tricky one. This file (`admin/modules/analytics.js`) isn't a typical [GOV.UK JS module](https://docs.publishing.service.gov.uk/repos/govuk_publishing_components/javascript-modules.html) because it automatically runs on page load, rather than exporting a Class/Prototype/Function that will later be instantiated on a DOM element.

That makes it pretty difficult to test since it will only be executed once, on page load, when the Jasmine test runner first loads up. Whereas the unit tests need to be able to instantiate Analytics multiple times over, to assert that the correct Custom Dimensions are set.

@DilwoarH is there a better / established pattern that can be used here instead of putting it in the `GOVUK` namespace?

Reply:
Our Jasmine test suite is only really suitable for unit testing JavaScript code. The ideal 'unit' of JavaScript would be an isolated function that can be called & torn down repeatedly. And since these unit tests all run within a single page in the browser, they shouldn't really rely on global state belonging to the overall page (e.g. `<meta>` tags present on the Jasmine test page).

> @DilwoarH: Is there a way to add the meta tag before the jasmine JS runs?

Therefore, this doesn't feel like a great option. It's kind of bending the unit test suite to do something it wasn't designed to do, which is why it feels a bit weird – it's the code telling us it's a bad fit for unit testing.

It feels like we have a couple of options:

1. **Unit test:** If we want to unit test this, we need the behaviour to be accessible from our Jasmine tests. This is the route @jyoung-gds has taken by 'exporting' that `setCustomDimensionsFromMetaTags` function to the `window.GOVUK` global object, so our unit tests can call it.
2. **Integration test:** If we want this to effectively be an 'anonymous' or private function which only runs once on page load, then it'd probably be more appropriate to cover its behaviour in a higher-level JS-enabled integration test. That way we'll be able to reload the page to easily test its behaviour under specific conditions.

Maybe we should be exploring Option 2 here. What do you think?